drpcsocket/stream.ts

```diff
@@ -0,0 +1,173 @@
+        super();
+
+        this.opts = opts || {splitSize: 0};
+
+        this.writeCount = 0;
+
+        this.id = new ID({
+            stream: id,
+            message: uint64.new(0),
+        });
+        this.writer = writer;
+
+        this.currentState = States.OPEN;
+    }
+
+    handlePacket(packet: Packet) {
+        if (this.currentState >= States.TERMINATED) {
+            // terminated / cancelled / finished
+            return;
+        }
+
+        switch (packet.kind) {
+            case Kind.INVOKE:
+                let err = new ProtocolError("invoke on existing stream");
+                this.end(err);
+                return;
+
+            case Kind.MESSAGE:
+                if (this.currentState == States.RECV_CLOSED) {
+                    return;
+                }
+
+                this.emit("message", packet.data);
+                return;
+
+            case Kind.ERROR:
+                this.end(parseError(packet.data));
+                return;
+
+            case Kind.CLOSE:
+                this.end(endClosed);
+                return;
+
+            case Kind.CLOSE_SEND:
+                if (this.currentState == States.SEND_CLOSED) {
+                    this.end(endBothClosed);
+                } else {
+                    this.currentState = States.RECV_CLOSED;
+                }
+
+                return;
+
+            default:
+                this.end(new InternalError(`unknown packet kind ${packet.kind.toString()}`));
+        }
+    }
+
+    //
+    // helpers
+    //
+
+    private newPacket(kind: Kind, data: Buffer) {
+        this.id.message = this.id.message.add(1);
+
+        return new Packet({
+            data,
+            id: new ID({
+                stream: this.id.stream,
+                message: this.id.message,
+            }),
+            kind,
+        });
+    }
+
+    private checkFinished() {
+        if (this.currentState == States.TERMINATED && this.writeCount == 0) {
+            this.currentState = States.FINISHED;
+            this.emit("close", this.err);
+        }
+    }
+
+    end(err?: Error) {
+        this.currentState = States.TERMINATED;
+        this.err = this.err ? this.err : err;
+        this.checkFinished();
+    }
+
+    //
+    // raw read/write
+    //
+
+    async write(kind: Kind, data: Buffer): Promise<void> {
+        if (this.currentState >= States.TERMINATED) {
+            if (this.err) {
+                throw this.err;
+            }
+
+            // terminated / cancelled / finished
+            return;
+        } else if (this.currentState == States.SEND_CLOSED) {
+            throw sendClosed;
+        }
+
+        this.writeCount++;
+        try {
+            let promises: Array<Promise<void>> = [];
+
+            split(this.newPacket(kind, data), this.opts.splitSize, (frame: Frame) => {
+                promises.push(this.writer.writeFrame(frame));
+            });
+
+            await Promise.all(promises);
```

Comment:
I'm not very familiar with Promise semantics. Is it possible for this block of code to send the frames out of order? Like if a packet splits into frames 1 and 2, can the receiving end observe 2 and then 1?

Reply:
Good question.... I think it might depend on the underlying writable implementation... I should be able to chain the writes pretty easily to ensure one is sent after another, etc...

Reply:
If they have to be sent in order a for-of might make sense as well

Reply:
it's pretty easy to do `chain = chain.then(() => ....)`